### PR TITLE
WIP byte string parser input for request target parsing

### PR DESCRIPTION
**Pull request by johanandren**
Based on the prefused branch for now, will rebase when that is merged.

Local tests showed ~5% throughput increase. Serious benchmarking needed to confirm.

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Note that this ParserInput type will NOT work with general `UTF-8`-encoded input as this can contain
+ * character representations spanning multiple bytes. However, if you know that your input will only ever contain
+ * 7-bit ASCII characters (0x00-0x7F) then UTF-8 is fine, since the first 127 UTF-8 characters are
+ * encoded with only one byte that is identical to 7-bit ASCII and ISO-8859-1.
+ */
+final class ByteStringParserInput(bytes: ByteString) extends DefaultParserInput {
+  override def charAt(ix: Int): Char = (bytes(ix) & 0xFF).toChar
+  override def length: Int = bytes.size
+  override def sliceString(start: Int, end: Int): String = bytes.slice(start, end - start).decodeString(StandardCharsets.ISO_8859_1)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala]: that's the charset the previous one used too right?

Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala]: Yeah, but it does it like this: ```new String(bytes, start, end - start, `ISO-8859-1`)```

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/ByteStringParserInput.scala]: Ok I see

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala [hunk 2]
```diff
@@ -14,160 +15,197 @@ import akka.http.scaladsl.model._
...
+      }
+    }
+
+    override def parseMessage(input: ByteString, offset: Int): StateResult = {
+      var cursor = parseMethod(input, offset)
+      cursor = parseRequestTarget(input, cursor)
+      cursor = parseProtocol(input, cursor)
+      if (byteChar(input, cursor) == '\r' && byteChar(input, cursor + 1) == '\n')
+        parseHeaderLines(input, cursor + 2)
+      else onBadProtocol
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala]: `onBadProtocol()` would be nicer, it throws

File: akka-http-core/src/test/scala/akka/http/scaladsl/TestServer.scala [hunk 3]
```diff
@@ -23,10 +23,17 @@ object TestServer extends App {
...
     akka.stream.materializer.debug.fuzzing-mode = off
+    akka.actor.serialize-creators = off
+    akka.actor.serialize-messages = off
+    akka.actor.default-dispatcher.throughput = 1000
     """)
   implicit val system = ActorSystem("ServerTest", testConf)
-  implicit val fm = ActorMaterializer()
 
+  val settings = ActorMaterializerSettings(system)
+    .withFuzzing(false)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/TestServer.scala]: no need for this becayse `     akka.stream.materializer.debug.fuzzing-mode = off
`

File: akka-http-core/src/test/scala/akka/http/scaladsl/TestClient.scala [hunk 4]
```diff
@@ -62,5 +71,47 @@ object TestClient extends App {
...
+              .map(res ⇒ (url, path, res)): Future[(String, Path, IOResult)]
+
+            headers.flatMap(_ ⇒ body)
+          }
+      }
+
+      println("Fetched urls: " + Await.result(done, 10.minutes))
+    } finally {
+      Http().shutdownAllConnectionPools().flatMap(_ ⇒ system.terminate())
+    }
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/TestClient.scala]: ok, why not

File: akka-http-core/src/test/scala/akka/http/impl/util/ByteStringParserInputSpec.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package akka.http.impl.util
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/util/ByteStringParserInputSpec.scala]: copyright header

File: akka-http-core/src/test/scala/akka/http/impl/engine/parsing/RequestParserSpec.scala [hunk 6]
```diff
@@ -557,7 +557,7 @@ class RequestParserSpec extends FreeSpec with Matchers with BeforeAndAfterAll {
     def multiParse(parser: HttpRequestParser)(input: Seq[String]): Seq[Either[RequestOutput, StrictEqualHttpRequest]] =
       Source(input.toList)
         .map(bytes ⇒ SessionBytes(TLSPlacebo.dummySession, ByteString(bytes)))
-        .via(parser.stage).named("parser")
+        .via(parser).named("parser")
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/RequestParserSpec.scala]: 👍

**Comment by ktoso** [comment]
LGTM, let's measure on our boxes

**Comment by johanandren** [comment]
Didn't rebase cleanly, continued in #21018

**Pull request closed without merging**